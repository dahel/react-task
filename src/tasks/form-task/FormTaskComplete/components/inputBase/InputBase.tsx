import { SyntheticEvent } from 'react';
import styled from 'styled-components';

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  min-width: 200px;
  height: 60px;
`;

const Label = styled.label`
  padding: 0 0 5px 5px;
`;

const Input = styled.input`
  padding: 5px;
  border: solid 1px ${({ error }: { error: boolean }) => (error ? 'red' : '#BDBDBD')};
  border-radius: 4px;
`;

const Required = styled.span`
  &::after {
    content: '*';
    color: red;
  }
`;

const RequiredMessage = styled.span`
  color: red;
  font-size: 11px;
  height: 13px;
  text-indent: 5px;
  margin-top: 1px;
`;

interface IProps {
  value: string;
  label: string;
  name: string;
  placeholder?: string;
  onBlur?: (event: SyntheticEvent) => void;
  onChange?: (event: SyntheticEvent) => void;
  required?: boolean;
  type?: string;
  validationError?: string;
  valid?: boolean;
  disabled?: boolean;
}

const InputBase = ({
  value,
  label,
  name,
  placeholder = '',
  onBlur = () => {},
  onChange = () => {},
  required = false,
  type = 'text',
  validationError = '',
  valid = true,
  disabled = false,
}: IProps) => {
  return (
    <Wrapper>
      <Label htmlFor={name}>
        {label} {required && <Required />}
      </Label>
      <Input
        disabled={disabled}
        value={value}
        onChange={onChange}
        onBlur={onBlur}
        id={name}
        type={type}
        name={name}
        placeholder={placeholder}
        error={!valid}
      />
      {!valid && <RequiredMessage>{validationError}</RequiredMessage>}
    </Wrapper>
  );
};

export default InputBase;
