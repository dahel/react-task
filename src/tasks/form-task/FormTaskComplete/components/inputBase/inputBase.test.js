import renderer from 'react-test-renderer';
import InputBase from './InputBase';

describe('InputBase component', () => {
  const props = {
    name: 'name',
    label: 'label',
    value: 'value',
  };

  test('renders properly for only-required props passed', () => {
    const component = renderer.create(<InputBase {...props} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('renders properly for given placeholder', () => {
    const component = renderer.create(<InputBase {...props} placeholder="placeholder" />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('renders properly for given type', () => {
    const component = renderer.create(<InputBase {...props} type="number" />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test(`renders properly in case 'disabled' prop is passed`, () => {
    const component = renderer.create(<InputBase {...props} disabled={true} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test(`renders properly in case 'disabled' prop is passed`, () => {
    const component = renderer.create(<InputBase {...props} disabled={true} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test(`renders properly in case 'valid' prop is false and 'validationError' prop is defined`, () => {
    const component = renderer.create(<InputBase {...props} valid={false} validationError="bad value" />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test(`renders properly in case 'valid' prop is false and 'validationError' is not defined`, () => {
    const component = renderer.create(<InputBase {...props} valid={false} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test(`calls 'onChange' callback passed as prop`, () => {
    const onChange = jest.fn();
    const root = renderer.create(<InputBase {...props} onChange={onChange} />).root;
    const input = root.findByType('input');

    input.props.onChange({ name: 'change' });

    expect(onChange).toHaveBeenCalledWith({ name: 'change' });
  });

  test(`calls 'onBlur' callback passed as prop`, () => {
    const onBlur = jest.fn();
    const root = renderer.create(<InputBase {...props} onBlur={onBlur} />).root;
    const input = root.findByType('input');

    input.props.onBlur({ name: 'blur' });

    expect(onBlur).toHaveBeenCalledWith({ name: 'blur' });
  });
});
