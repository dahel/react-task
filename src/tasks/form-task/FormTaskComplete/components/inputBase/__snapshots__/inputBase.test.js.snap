// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputBase component renders properly for given placeholder 1`] = `
<div
  className="sc-bdvvtL dIIVTQ"
>
  <label
    className="sc-gsDKAQ cSZgOM"
    htmlFor="name"
  >
    label
     
  </label>
  <input
    className="sc-dkPtRN fVDXDy"
    disabled={false}
    id="name"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="placeholder"
    type="text"
    value="value"
  />
</div>
`;

exports[`InputBase component renders properly for given type 1`] = `
<div
  className="sc-bdvvtL dIIVTQ"
>
  <label
    className="sc-gsDKAQ cSZgOM"
    htmlFor="name"
  >
    label
     
  </label>
  <input
    className="sc-dkPtRN fVDXDy"
    disabled={false}
    id="name"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="number"
    value="value"
  />
</div>
`;

exports[`InputBase component renders properly for only-required props passed 1`] = `
<div
  className="sc-bdvvtL dIIVTQ"
>
  <label
    className="sc-gsDKAQ cSZgOM"
    htmlFor="name"
  >
    label
     
  </label>
  <input
    className="sc-dkPtRN fVDXDy"
    disabled={false}
    id="name"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
    value="value"
  />
</div>
`;

exports[`InputBase component renders properly in case 'disabled' prop is passed 1`] = `
<div
  className="sc-bdvvtL dIIVTQ"
>
  <label
    className="sc-gsDKAQ cSZgOM"
    htmlFor="name"
  >
    label
     
  </label>
  <input
    className="sc-dkPtRN fVDXDy"
    disabled={true}
    id="name"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
    value="value"
  />
</div>
`;

exports[`InputBase component renders properly in case 'disabled' prop is passed 2`] = `
<div
  className="sc-bdvvtL dIIVTQ"
>
  <label
    className="sc-gsDKAQ cSZgOM"
    htmlFor="name"
  >
    label
     
  </label>
  <input
    className="sc-dkPtRN fVDXDy"
    disabled={true}
    id="name"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
    value="value"
  />
</div>
`;

exports[`InputBase component renders properly in case 'valid' prop is false and 'validationError' is not defined 1`] = `
<div
  className="sc-bdvvtL dIIVTQ"
>
  <label
    className="sc-gsDKAQ cSZgOM"
    htmlFor="name"
  >
    label
     
  </label>
  <input
    className="sc-dkPtRN fFQvcc"
    disabled={false}
    id="name"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
    value="value"
  />
  <span
    className="sc-eCImPb bZaMKe"
  >
    
  </span>
</div>
`;

exports[`InputBase component renders properly in case 'valid' prop is false and 'validationError' prop is defined 1`] = `
<div
  className="sc-bdvvtL dIIVTQ"
>
  <label
    className="sc-gsDKAQ cSZgOM"
    htmlFor="name"
  >
    label
     
  </label>
  <input
    className="sc-dkPtRN fFQvcc"
    disabled={false}
    id="name"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
    value="value"
  />
  <span
    className="sc-eCImPb bZaMKe"
  >
    bad value
  </span>
</div>
`;
